### ISSUE TITLE
Support to cleanup obsolete indices for V3 segment format

### ISSUE THREAD

**Issue by klsince** [apache/pinot#7320]
For V3 segment format, currently only Startree index can be removed from disk on Pinot servers. Other indices, like inverted index, json index, text index are still kept at server side even if they are removed from table config. This issue proposes to clean up all kinds of indices for V3 segment format, if they are deleted from table config.

**Comment by klsince** [comment]
related PR: https://github.com/apache/pinot/issues/7320

**Comment by klsince** [comment]
Related: https://github.com/apache/pinot/pull/7294

**Comment by klsince** [comment]
Related: https://github.com/apache/pinot/pull/7297

**Comment by mcvsubbu** [comment]
Even if indices are on disk, they are not loaded if they are not turned on. So, why bother with this? Am I missing something?

**Comment by klsince** [comment]
yeah, but they still occupy the disk space, and we recently saw an incident as disk usage went up high due to those obsolete indices. Though there are ways to mitigate the incident, but as prevention, we planned to clean up indices when they are deleted from table config.

**Comment by mcvsubbu** [comment]
@kishoreg, @siddharthteotia  what do you think? Does this warrant a change in pinot code? We do allow re-push of segments.

**Comment by mayankshriv** [comment]
+1 on cleaning obsolete indices from sever. This comes in handy in cases where users are trying to optimize indices and may try out different options. While the obsolete indices themselves do not cause any issues with CPU/Memory usage of Pinot servers, they may eat up disk storage that can be expensive, moreover, may prevent pushing of more data (experienced in certain deployments already). This can be achieved in multiple ways:
- Re-download of segments from controller/deep-store, and then build the necessary indices as per the indexing config.
- Let the servers themselves cleanup the indices (as part of reload) locally, without needing to download from controller.

**Comment by klsince** [comment]
Is 're-push of segment' the feature described in this issue https://github.com/apache/pinot/issues/7321? 

In fact, as followups from the incident, we planned these two options: 1) add option to force segment download (mainly for mitigation); 2) enable index cleanup for V3 format, so that indices on servers stay in sync with what's set in table config not just when adding new ones but also when removing old ones.

In general, re-push of segment is kinda heavy lifting for index cleanup particularly for large table, as it incurs considerable network egress cost while downloading large volume of data. Besides, letting indices on servers stay in sync with what's set in table config makes it a bit more intuitive to use Pinot, basically deleting indices would reduce disk usage.

**Comment by mcvsubbu** [comment]
@klsince thanks for connecting the dots. I see that the feature can be useful. Mayank has proposed two approaches to do this. A third (less disruptive code-wise) one is to have minions cleanup the segments via a background task.

Let us discuss a bit on these before you start with the PRs.

Which approach are you choosing and why?

**Comment by klsince** [comment]
Thanks for the 3rd proposal!

I'm doing both option 1 (force download) and 2 (server local cleanup) currently. Option 1 is more like one last resort to bring things back to life, not necessarily just for index cleanup. 

Option 2 is to clean indices more efficiently and supposed to happen more frequently; and as it's done during segment reloading, it can leverage the existing reloading mechanism to be atomic, thus transparent to queries on the fly.

Option 3 - using Minion - I don't see it conflicts with the options above. When minion tasks are combing the segments, they can do cleanup, and also things like merging and rollup as proposed in another project. But as it's happening async and potentially outside server hosts, it might be more costly due to data transfer and non-trivial to swap in the new segment transparently to queries on the fly. 

Thoughts?

**Comment by Jackie-Jiang** [comment]
@mcvsubbu We want to support all 3 approaches, and each of them has their own pros/cons

- Force download: Big hammer solution, can be used to fix corrupted segment on server, but costly
- Local cleanup: Cheap, align with the current way of adding/removing indexes on-the-fly when server loads the segment. We don't remove the index right now only because V3 format does not support it. This issue ticket is trying to resolve that.
- Minion cleanup: Powerful, can perform operations that cannot be done locally, e.g. dictionary encoded to raw index conversion, but costly, and requires setting up minion instances

@siddharthteotia and I have been discussing the V3 removal several times, but haven't got a chance to work on it

**Comment by mcvsubbu** [comment]
Why support all three? 
(1) Corrupted segment (should be few) can be fixed by just deleting the segment from the servrs, that will trigger a download on restart or enable/disable API.
(2) Local cleanup needs to be thought through a bit. If you are doing this, please write down your approach. Are you doing this in a background thread? Are you doing it at segment load time? (load could take longer). Please write up something about this before we proceed further.
(3) Minions are meant to do this. We need to add cpu's to do the re-build of segment _some_ place. Why not minions? It can be clean, enabled in installations that need it, and can also be manually triggered.

**Comment by mayankshriv** [comment]
Side note for 3) Deleting of indices should not be CPU intensive.

**Comment by klsince** [comment]
Re 2, local index cleanup works as part of segment reloading. The cleanup happens in the same thread doing segment reloading, and it leverages the current failure handling mechanism (as in [code](https://github.com/apache/pinot/blob/master/pinot-server/src/main/java/org/apache/pinot/server/starter/helix/HelixInstanceDataManager.java#L276)) of segment reloading to keep disk states consistent upon failures, and be atomic when swapping in the cleaned segment. With this atomicity, the queries can continue to work with the existing segment until the new one is swapped in.

Just like adding new indices during segment reloading, removing indices is also done inside the segment folder **not** accessed by ongoing queries, so it's safe to modify the files in the folder. To cleanup, we simply copy the indices defined in table config into a temp file, then rename it back to index file, effectively removing those not set in table config any more. In implementation, we use transferTo() for copy as in [the PR](https://github.com/apache/pinot/pull/7301/files#diff-52f126f7138a706a5fddb9257af1c558c4623269bc69308212a77c06021cbef7R433)

Besides, cleanup happens after adding new indices so that newly created indices are kept after cleanup. In implementation, it just needs to do cleanup after closing all PinotDataBuffers as in [the PR](https://github.com/apache/pinot/pull/7301/files#diff-52f126f7138a706a5fddb9257af1c558c4623269bc69308212a77c06021cbef7R365). 

Hope this helps clarify a bit more, and feel free to comment the PR. Thanks!

Re 1, I assume it's not always available to ssh to servers to delete segments, so an option to force segment download via Pinot restful API can be convenient.
